mautic/mautic
# Generated columns

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | Y
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | /
| Related developer documentation PR URL | /
| Issues addressed (#s or URLs) | /
| BC breaks? | N
| Deprecations? | N

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR is adding a [generated column](https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html) based on `email_stats.date_sent`. It will generate the value in format `Y-m-d` and build a new index from it. Then it modifies the line chart queries to use that column instead of formatting the original date.

Original query to load line chart data from email stats table:
```
SELECT DATE_FORMAT(t.date_sent, '%Y-%m-%d') AS date, COUNT(*) AS count FROM mautic_email_stats t WHERE (t.email_id IN (3)) AND (t.date_sent BETWEEN :dateFrom AND :dateTo) GROUP BY DATE_FORMAT(t.date_sent, '%Y-%m-%d') ORDER BY DATE_FORMAT(t.date_sent, '%Y-%m-%d') ASC LIMIT 31
```

New query based on the generated column:
```
SELECT t.generated_sent_date AS date, COUNT(*) AS count FROM mautic_email_stats t WHERE (t.email_id IN (3)) AND (t.date_sent BETWEEN :dateFrom AND :dateTo) GROUP BY t.generated_sent_date ORDER BY t.generated_sent_date ASC LIMIT 31
```

The speed improvement for the select query is somewhere around 3,5x faster. To be more specific, running the query on that 6.6M row table took ~700ms. With this PR it loads in ~200ms.

I've tested adding the new generated column on a email_stats table with 6.6M rows and it took 172 seconds to add it. So it should not cause any troubles to migrate even big instances. But do not use the UI updater. Use the script!

This is an experiment. If we find that it's not causing any troubles we should implement it also for date_read and other date columns in other stat tables used in the line charts.

#### Steps to test this PR:
1. Ensure you can see the generated column definition when viewing the schema changes. You should see something like this:
```
$ app/console doctrine:schema:update --dump-sql | grep generated
ALTER TABLE mautic_email_stats ADD generated_sent_date DATE AS (CONCAT(YEAR(date_sent), "-", LPAD(MONTH(date_sent), 2, "0"), "-", LPAD(DAY(date_sent), 2, "0"))) COMMENT '(DC2Type:generated)';
CREATE INDEX mautic_generated_sent_date_email_id ON mautic_email_stats (generated_sent_date, email_id);
```
2. Run the migration. Generated fields doesn't have a migration file because we don't know whether the database supports generated fields. So we check the database version and run the alter table queries only if the database supports it. You should see this during migration:
```
$ app/console doctrine:migration:migrate

++ adding generated column generated_sent_date
-> ALTER TABLE mautic_email_stats
            ADD generated_sent_date DATE AS (CONCAT(YEAR(date_sent), "-", LPAD(MONTH(date_sent), 2, "0"), "-", LPAD(DAY(date_sent), 2, "0"))) COMMENT '(DC2Type:generated)',
            ADD index `mautic_generated_sent_date_email_id`(generated_sent_date, email_id)
++ generated column added (generated columns: 60.00 MiB - 36 ms)
                    Mautic Migrations
WARNING! You are about to execute a database migration that could result in schema changes and data lost. Are you sure you wish to continue? (y/n)No migrations to execute.
```

3. Load a segment email detail page, report based on email stats with the line chart and dashboard with the emails sent widget. All should load correctly (hopefully faster than before) and it should show the same values as before. Test also the sent email dashboard widget and report widget. All time units should still work.
